Updated dependencies for QCG-PJ

mzrghorbani opened the pull request:
## Features Added

- QCG-PilotJob support via PJ_TYPE=QCG argument in run_ensemble()
- New job preparation flow for QCG:
  - Tasks are mapped to .sh scripts based on sweep inputs
  - Each task is converted to a QCG job description via template

- Support for replicated execution:
  - Users can specify upsample="d1;d2" and replica_start_number="2;2"
  - Replicated directories are created (e.g. d1/replica1/, d1/replica2/)

- Clean handling of Python environments:
  - Optional use of virtual_env_path from machines_user.yml
  - Fallback to active virtual environment or user-defined Python

- Automatic remote staging:
  - SWEEP contents and run scripts are rsynced to the remote machine
  - Ensures tasks are available before QCG execution

- QCG task script (qcg_manager_<job>.py) with:
  - Logging
  - Resource detection
  - Runtime diagnostics
  - Graceful failure handling

## Usage Example

```bash
fabsim archer2 dummy_ensemble:dummy_test,PJ=True,PJ_TYPE=QCG,venv=true
```

Additional test required for:

```bash
fabsim archer2 dummy_ensemble:dummy_test,upsample="d1;d2",replica_start_number="2;2",PJ=True,PJ_TYPE=QCG,venv=true
```

**Please do not merge until the features are fully tested.**

lauraharbach commented:
I have tested the two suggested examples, and they appear to produce the correct dummy results as expected. However, there seems to be an issue with the `fabsim archer2 fetch_results` command. While the results are successfully generated and stored on archer2, they are not being synced locally via rsync. Could you please investigate this issue? Let me know if you need additional details or logs to debug further.

lauraharbach commented:
Should there be a replica_end_number to complement the replica_start_number? Currently, it seems to run only replica2 for each execution. Is this the intended behavior, or should it handle a range of replicas instead?

lauraharbach commented:
Would it be better to move the TaskID3_XXX.stderr/out files to a separate directory? Keeping them in the directory for each run/replica makes it a bit cumbersome to manage and analyse the results, especially for SA.

lauraharbach commented on fabsim/base/fab.py:
```diff
@@ -47,11 +47,7 @@ def get_plugin_path(name: str) -> str:
     """
     plugin_path = os.path.join(env.localroot, "plugins", name)
     if not os.path.exists(plugin_path):
-        raise RuntimeError(
```
I think the original error message was more comprehensive and useful.

lauraharbach commented on fabsim/deploy/templates/qcg-PJ-py:
```diff
@@ -1,27 +1,92 @@
+    # Clean up
+    logging.info("All jobs completed, shutting down QCG-PilotJob manager")
+    try:
+        manager.finish()
```
Does this need to be to maintain original functionality?
manager.finish()
manager.stopManager()
manager.cleanup()

mzrghorbani commented:
I'm closing this pull request as it was created primarily for testing purposes and did not pass the GitHub workflow tests. 

I've since created a new branch (qcg-dev) with a more comprehensive approach to QCG-PilotJob integration that passes all tests successfully. The new PR includes:

- Direct installation method for QCG-PilotJob
- Virtual environment management
- Fixed task description generation
- Updated documentation
- Style fixes

Please see PR #319 for the complete and working implementation. Thank you for your comments.
